common/crypto/attestation-api/attestation/attestation.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "attestation.h"
+#include <string>
+#include "attestation_tags.h"
+#include "base64.h"
+#include "error.h"
+#include "logging.h"
+#include "parson.h"
+#include "pdo/common/crypto/crypto.h"
+#include "pdo/common/jsonvalue.h"
+#include "sgx_quote.h"
+#include "sgx_utils.h"
+#include "types.h"
+
+/**********************************************************************************************************************
+ * C prototype declarations for the ocalls
+ * *******************************************************************************************************************/
+#ifdef __cplusplus
+extern "C" {
+#endif
+sgx_status_t ocall_init_quote(
+    uint8_t* target, uint32_t target_len, uint8_t* egid, uint32_t egid_len);
+sgx_status_t ocall_get_quote(uint8_t* spid,
+    uint32_t spid_len,
+    uint8_t* sig_rl,
+    uint32_t sig_rl_len,
+    uint32_t sign_type,
+    uint8_t* report,
+    uint32_t report_len,
+    uint8_t* quote,
+    uint32_t max_quote_len,
+    uint32_t* actual_quote_len);
```

Comment:
> However, i guess to make this really modular, we would have to treat the different edl's separatedly via edger8 ather than including them (recursively) into a single .edl as otherwise you will end up with a single ..._t.h

AFAIK, that's not possible, and, yes, you will end up with a single .h (and a single .c).
Also consider that this method is also used by the sdk itself -- e.g., see `sgx_tstdc.edl`.
The point is that there is a single table for ecalls/ocalls that the edger8r creates.

> a single ..._t.h which you really cannot include here as from a module perspective you don't even have an idea that what other code includes you.

I don't need to include it here. Also, in general, you need the function prototype, not necessarily the header. See the `common/logging.c` in this PR!!

> And doing everything separate has other drawbacks besides all the refactoring of build scripts as you loose the ability to see at a single place the enclave interface. All that said, might be worth to add a comment here that we have to be careful that this interface is identical to the edger8r generated interface?

I answered before. It's already intrinsic in the sdk when you use the other edls. So, if you want the final complete interface, look at the edger8r-generated files.

Reply:
> The point is that there is a single table for ecalls/ocalls that the edger8r creates.

Good point: you need a single edl exactly because of this table
 
> > a single ..._t.h which you really cannot include here as from a module perspective you don't even have an idea that what other code includes you.
> 
> I don't need to include it here. Also, in general, you need the function prototype, not necessarily the header. See the `common/logging.c` in this PR!!

Well, my point is that edger8r already automatically defines the correct function prototype (see `ecc_enclave/enclave/enclave_t.h`), that's why i first was puzzled why we redefine it here.   But from a modularity perspective there is no good way to include that header-file.  Seems to me actually an SGX SDK issue (e.g., they could generate a .h for each .edl separately, with the header-files just including sub-header-files via "#include; in that case you could have actually included attestation_t.h or alike without having to duplicate the function prototype ..)

Reply:
Sort of a chicken-n-egg problem.

Your point is accurate.
What I would question is the need/appropriateness of the edger8r to (i) generate those prototypes and (ii) the way it does it. It's usually convenient, but the edger8r might be parameterized for these cases.

Reply:
hmm, the edl is the reference of the API and implies the prototype, so to me edger8r is exactly what should define that prototype as it can be automatically generated. Why would you want to manually derive what you can do automatically? Anyway, discussion goes way beyond this PR ... :-)